File: src/main/java/io/github/thejaxonhill/tarotapi/exception/ExceptionHandlers.java
package io.github.thejaxonhill.tarotapi.exception;

import java.io.IOException;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class ExceptionHandlers {

    @ExceptionHandler(value = { RuntimeException.class })
    protected ResponseEntity<Object> handleRuntimeException(RuntimeException ex) {
        return ResponseEntity.badRequest().body(ex.getMessage());
    }

    @ExceptionHandler(value = { IOException.class })
    protected ResponseEntity<Object> handleIOException(RuntimeException ex) {
        return ResponseEntity.badRequest().body(ex.getMessage());
    }

}

File: src/main/kotlin/io/github/thejaxonhill/tarotapi/infrastructure/ExceptionHandlers.kt
package io.github.thejaxonhill.tarotapi.infrastructure

import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

@ControllerAdvice
class ExceptionHandlers {
    @ExceptionHandler(value = [RuntimeException::class])
    fun handleRuntimeException(ex: RuntimeException): ResponseEntity<ErrorResponse> {
        return ResponseEntity.badRequest().body(ex.toErrorResponse())
    }

    @ExceptionHandler(value = [Exception::class])
    fun handleIOException(ex: Exception): ResponseEntity<ErrorResponse> {
        return ResponseEntity.internalServerError().body(ex.toErrorResponse())
    }
}

data class ErrorResponse(val message: String? = null, val localizedMessage: String? = null)

private fun Exception.toErrorResponse(): ErrorResponse = ErrorResponse(message, localizedMessage)